eth/downloader/downloader.go

```diff
@@ -1352,6 +1322,11 @@ func (d *Downloader) fetchNodeData() error {
 		deliver = func(packet dataPack) (int, error) {
 			start := time.Now()
 			return d.queue.DeliverNodeData(packet.PeerId(), packet.(*statePack).states, func(err error, delivered int) {
+				// If the peer gave us nothing, stalling fast sync, drop
+				if delivered == 0 {
```

Comment:
I'm not sure that dropping in these circumstances is the right behaviour. What if a node has throttling build in and decides not to give you the data because it's already overburdened with request and decides to send you an empty response (instead of ignoring it).

I can imagine something like this happening when we've got light nodes.